### PR TITLE
rework applying closure requirements in borrowck

### LINKED ISSUE #99245: [NLL] loss of span for type outlives errors in closures/async blocks
Given the following code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=aaf294b9353da9ea4ef76206ed0c0a76

```rust
fn test<'a, T>() {
    || {
        None::<&'a T>;
    };
}
```

The current output is:

```
error[E0309]: the parameter type `T` may not live long enough
 --> src/lib.rs:2:5
  |
2 | /     || {
3 | |         None::<&'a T>;
4 | |     };
  | |_____^ ...so that the type `T` will meet its required lifetime bounds
  |
help: consider adding an explicit lifetime bound...
  |
1 | fn test<'a, T: 'a>() {
  |              ++++
```

Ideally the output should look similar to when constraint is generated in the main fn body:

```
error[E0309]: the parameter type `T` may not live long enough
 --> src/lib.rs:3:9
  |
3 |         None::<&'a T>;
  |         ^^^^^^^^^^^^^ ...so that the type `T` will meet its required lifetime bounds
  |
help: consider adding an explicit lifetime bound...
  |
1 | fn test<'a, T: 'a>() {
  |              ++++
```

@rustbot label A-NLL NLL-diagnostics D-terse regression-from-stable-to-beta

### PR DESCRIPTION
**Pull request by aliemjay**
Previously the promoted closure constraints were registered under the category `ConstraintCategory::ClosureBounds` in `type_check::prove_closure_bounds()` and then mapped back their original category in `regions_infer::best_blame_constraint` using the complicated map `closure_bounds_mapping`.

Now we're registering promoted constraints under their original category and span earlier in `type_check::prove_closure_bounds`.

See commit messages.

Fixes #99245

### CODE REVIEW

File: src/test/ui/generic-associated-types/issue-91139.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+//~ ERROR the parameter type `T` may not live long enough
```
Comment (cjgillot) [inline src/test/ui/generic-associated-types/issue-91139.rs:1]: This message is emitted with `DUMMY_SP`. This is unfortunate. Any idea where it comes from?

Comment (aliemjay) [reply src/test/ui/generic-associated-types/issue-91139.rs:1]: yeah, I forgot to comment on this one. This indeed comes from https://github.com/rust-lang/rust/blob/37928f59865b5ff53b95fe2f3c02491890a4129d/compiler/rustc_borrowck/src/type_check/free_region_relations.rs#L305
which I've fixed in #99818 to make it point to the closure body instead.

File: src/test/ui/nll/ty-outlives/projection-two-region-trait-bound-closure.stderr [hunk 2]
```diff
@@ -167,17 +167,20 @@ LL | |     T: Anything<'b, 'b>,
...
+  --> $DIR/projection-two-region-trait-bound-closure.rs:87:5
    |
 LL | fn two_regions<'a, 'b, T>(cell: Cell<&'a ()>, t: T)
    |                --  -- lifetime `'b` defined here
    |                |
    |                lifetime `'a` defined here
 ...
 LL |     with_signature(cell, t, |cell, t| require(cell, t));
-   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^ closure body requires that `'b` must outlive `'a`
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ argument requires that `'b` must outlive `'a`
```
Comment (aliemjay) [inline src/test/ui/nll/ty-outlives/projection-two-region-trait-bound-closure.stderr:178]: This is another minor diagnostic regression, but I believe it never was an intended behavior of the previous setup. Actually at the point of rendering diagnostics the intention was to not have a constraint with `ClosureBounds` category, but here we've somehow missed converting `ClosureBounds` constraints.

Anyway this regression only affect the niche case where there is double lifetime bounds in assoc type definition (`Trait<'a, 'b>::Assoc: 'a + 'b`) AND the bound `X: Trait<'a, 'a>` appears in param env.

**Comment by aliemjay** [comment]
let me  spin the wheel again.
r?

**Comment by compiler-errors** [comment]
This change makes sense to me, and the second commit does have some good diagnostics improvements. r=me after rebasing.

**Comment by aliemjay** [comment]
Ah no bors delegatet command, then ping @compiler-errors to review.

**Pull request merged**